qdglue/metrics/grid_based_metrics.py

```diff
@@ -0,0 +1,290 @@
+import dataclasses
+from abc import ABC, abstractmethod
+from typing import Dict, Tuple
+
+import gymnasium
+import numpy as np
+
+from qdglue.types import Feature, Fitness
+
+
+@dataclasses.dataclass
+class DiscreteArchiveMetrics:
+    """Holds statistics about an archive.
+
+    Attributes:
+        coverage: Proportion of cells in the archive that have an elite - always
+            in the range $[0,1]$.
+        qd_score: QD score, i.e. sum of objective values of all elites in the
+            archive. All the fitnesses are normalized to $[0,1]$ using the
+            fitness_bounds before computing the QD score.
+        max_fitness: Maximum objective value of the elites in the archive.
+        ccdf: Complementary Cumulative Distribution of Fitness from Vassiliades
+            et al. (2018)
+    """
+
+    coverage: float
+    qd_score: float
+    max_fitness: float
+    ccdf: np.ndarray
```

Comment:
It seems the CCDF depends on what the edges of the bins are. I think this class should contain some way to reconstruct the bin edges -- for instance, it should have the edges themselves, or the fitness lower/upper bound with num_bins.

Reply:
I had those infos in the beginning, and removed them as I thought you would like the class to only contain the calculated metrics. So I added a `discrete_archive_metrics_functor` attribute (a bit ugly structure, I know, but that's working super well and it's super convenient).

Also added a `ccdf_fitness_bins` property that refinds the fitness bins using `discrete_archive_metrics_functor`.

Reply:
Another property that would be good to have is that we can save this class and reload it across versions of Python and even versions of the library. For instance, it would be nice if we could pickle this class (or at least save it as JSON) in version 0.1.0, and then in version 0.5.0 we're still able to open it.

One thing that would make this more difficult is storing a class type like `DiscreteArchiveMetricsCalculator`, which is likely to break across versions and thus not work with Pickle. I would prefer that we store info from the calculator class directly in this class; e.g., we can put the fitness bounds and ccdf fitness bins.

Regarding ccdf_fitness_bins, I think we can just store this directly on the class instead of creating the property. Keep in mind that:

- A property method gets re-evaluated every time you call it.
- The ccdf_fitness_bins are not necessarily evenly spaced, i.e., they may vary based on the calculator.

Reply:
I agree with most of your comments.
In particular, the `ccdf_fitness_bins` property was not an amazing idea.

[TO BE DISCUSSED]